src/main/java/com/microsoft/graph/tasks/PageIterator.java

```diff
@@ -0,0 +1,477 @@
+package com.microsoft.graph.tasks;
+
+import com.google.common.base.Strings;
+import com.microsoft.graph.CoreConstants;
+import com.microsoft.graph.exceptions.ServiceException;
+import com.microsoft.graph.requests.IBaseClient;
+import com.microsoft.kiota.HttpMethod;
+import com.microsoft.kiota.RequestAdapter;
+import com.microsoft.kiota.RequestInformation;
+import com.microsoft.kiota.serialization.AdditionalDataHolder;
+import com.microsoft.kiota.serialization.Parsable;
+import com.microsoft.kiota.serialization.ParsableFactory;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+/**
+ * A class for iterating through pages of a collection
+ * Uses to automatically pages through result sets across multiple calls and process each item in the result set.
+ * @param <TEntity> The type of the entity returned in the collection. This type must implement {@link Parsable}
+ * @param <TCollectionPage> The Microsoft Graph collection response type returned in the collection response. This type must implement {@link Parsable} and {@link AdditionalDataHolder}
+ */
+public class PageIterator<TEntity extends Parsable, TCollectionPage extends Parsable & AdditionalDataHolder> {
+    /**
+     * Creates a new instance of the PageIterator class
+     */
+    protected PageIterator() {
+        // default constructor
+    }
+    private static final String NO_COLLECTION_PROPERTY_ERROR = "The Parsable does not contain a collection property.";
+    private RequestAdapter requestAdapter;
+    private TCollectionPage currentPage;
+    private ParsableFactory<TCollectionPage> collectionPageFactory;
+    private Queue<TEntity> pageItemQueue;
+    private Function<TEntity, Boolean> processPageItemCallback;
+    private Function<TEntity, CompletableFuture<Boolean>> asyncProcessPageItemCallback;
+    private UnaryOperator<RequestInformation> requestConfigurator;
+
+
+    private String deltaLink;
+
+    /**
+     * The deltaLink returned from a delta query.
+     * @return the deltaLink from the delta query
+     */
+    @Nullable
+    public String getDeltaLink() {
+        return deltaLink;
+    }
+    private String nextLink;
+    /**
+     * The nextLink returned from a collection query.
+     * @return the nextLink from the collection query
+     */
+    @Nullable
+    public String getNextLink() {
+        return nextLink;
+    }
+    private PageIteratorState state = PageIteratorState.NOT_STARTED;
+    /**
+     * The state of the page iterator
+     * @return the state of the page iterator
+     */
+    @Nonnull
+    public PageIteratorState getPageIteratorState() {
```

Comment:
I wonder whether we should expose a getter for this state? It's a divergence from the spec. @andrueastman any opinion?

Reply:
I think it should be exposed and probably update the spec there. 
In the event that one is paging through a delta api, it may be useful to know if the return of the iteration is at the Delta state(a delta link is returned) and paging is paused due to reaching the end of the results. 
In this case the consumer can use the value to know that it should be safe to call `iterate()` again so that more delta results can be fetched.

Reply:
makes sense, @ramsessanchez do you feel up to updating the spec?  https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/PageIteratorTask.md

Reply:
Sure, would this go into the 'requirements' section?

Reply:
correct